### PR TITLE
NodeModel

### PR DESCRIPTION
**Pull request by NeginsCode**
Refactored NodeModel version which includes the functionality of MemoryNodeModel.
With the help of @josephbirkner it is now more pythonic. 

NodeModel gives you access to the labels, properties and relationships of your awesome nodes. 

Looking forward to your feedback.

                __
               / _)         
        .-^^^-/ /          
    __/       /              
    <__.|_|-|_|

### CODE REVIEW

File: src/scientio/node_model.py [hunk 1]
```diff
@@ -10,25 +11,51 @@ class RelationshipAvailability(Enum):
     NONE_AVAILABLE = 3
 
 
-class NodeModel(object):
-    def __init__(self, node=None):
-        if node is not None:
-            self.set(node)
+class NodeModel():
```
Comment (josephbirkner) [inline src/scientio/node_model.py]: Docs needed :robot:

File: src/scientio/node_model.py [hunk 2]
```diff
@@ -10,25 +11,51 @@ class RelationshipAvailability(Enum):
...
+
+    NAME = "name"
```
Comment (josephbirkner) [inline src/scientio/node_model.py]: Docs needed :robot:

File: src/scientio/node_model.py [hunk 3]
```diff
@@ -10,25 +11,51 @@ class RelationshipAvailability(Enum):
...
+    ID = "id"
+
+    def __init__(self, node=None, strip_query=None):
```
Comment (josephbirkner) [inline src/scientio/node_model.py]: Docs needed :robot:

File: src/scientio/node_model.py [hunk 4]
```diff
@@ -10,25 +11,51 @@ class RelationshipAvailability(Enum):
...
+
         # unique nodes ID assigned by memory
         self.id = 0
```
Comment (josephbirkner) [inline src/scientio/node_model.py]: Change `# ` comments to actual python docstrings :robot:

File: src/scientio/node_model.py [hunk 5]
```diff
@@ -10,25 +11,51 @@ class RelationshipAvailability(Enum):
...
         # Relation: <name, list od IDs>
-        self.relationships = dict()
+        self.relationships = defaultdict(list)
+
+        self.strip_query = False
+        if node is not None:
+            self.set_node(node)
+        self.strip_query = strip_query
+
+    def reset_id(self):
```
Comment (josephbirkner) [inline src/scientio/node_model.py]: Docs needed :robot:

File: src/scientio/node_model.py [hunk 6]
```diff
@@ -10,25 +11,51 @@ class RelationshipAvailability(Enum):
...
+        self.id = 0
+
+    def reset_label(self):
```
Comment (josephbirkner) [inline src/scientio/node_model.py]: Docs needed :robot:

**Comment by josephbirkner** [comment]
After consultation with @Waguramu yesterday:
- [ ] Please also remove the main label member.

**Comment by Waguramu** [comment]
Do not allow passing labels, name, relationships, property values in the constructor

Constructor should take either 
- a Type (ontology type to initialise a proper structure) 
- another Node (copy constructor)
- JsonNode

**Comment by Waguramu** [comment]
Add to_json() method

**Comment by josephbirkner** [comment]
> Add to_json() method

Can we postpone this? Would be cool if we could merge this PR asap imo.

File: src/scientio/node_model.py (hunk 6 above)
Comment (NeginsCode) [reply src/scientio/node_model.py]: Done :)

**Pull request merged**